### PR TITLE
Split the state tracking between the scanner and the parser.

### LINKED ISSUE #1374: Revisit state tracking in the parser vs scanner
For a better distribution of responsibilities, the state tracker needs some refactoring.

Token lookahead is necessary to determine the proper grammar production. This needs to exist in the parser as the scanner has no business regarding the production. The same goes for the extraneous location tracking for syntax node's location information.

The implementation of the reader algorithm for the pure tokenizer (i.e. `tokenize` API) shall not be in the parser as well. One day we will separate the tokenizer and find a new home for it, but for now it is sufficient to include it as part of the scanner.

### PR DESCRIPTION
**Pull request by ariya**
Token lookahead, extraneous location tracking for syntax node, etc are
mostly moved to the parser.

Reader algorithm (to determine whether `/` starts a regular expression
or not) for the pure tokenizer now resides completely within the
scanner.

Fixes #1374

### CODE REVIEW

File: src/scanner.ts [hunk 1]
```diff
@@ -1258,46 +1202,57 @@ export class Scanner {
...
+                case '}':
+                    // Dividing a function by anything makes little sense,
+                    // but we have to check for that.
+                    regex = false;
+                    if (this.tokenValues[this.openCurlyToken - 3] === 'function') {
+                        // Anonymous function, e.g. function(){} /42
+                        const check = this.tokenValues[this.openCurlyToken - 4];
+                        regex = check ? !beforeFunctionExpression(check) : false;
+                    } else if (this.tokenValues[this.openCurlyToken - 4] === 'function') {
+                        // Named function, e.g. function f(){} /42/
```
Comment (mikesherov) [inline src/scanner.ts]: Completely unrelated to this PR, but wanted to jot it down before I forgot....

Do we need to add cases for generator functions and arrow functions divided by a number here?

Comment (ariya) [inline src/scanner.ts]: > Do we need to add cases for generator functions and arrow functions divided by a number here?

Most like the reader algorithm needs an update anyway for ES6. Let's watch sweet.js and when it happens, also adopt the updated algorithm.

**Comment by ariya** [comment]
This is going to land soon so we can continue for further refactoring (yay!).

**Comment by mikesherov** [comment]
:+1:

**Pull request closed without merging**